File: docs/snippets/cas/overview/rules_factor_out.py
from mathy.core.rules import DistributiveFactorOutRule
from mathy.core.parser import ExpressionParser

input = "4x + 2x"
output = "(4 + 2) * x"
parser = ExpressionParser()

input_exp = parser.parse(input)
output_exp = parser.parse(output)

# Verify that the rule transforms the tree as expected
change = DistributiveFactorOutRule().apply_to(input_exp)
assert str(change.result) == output

# Verify that both tress evaluate to the same value
ctx = {"x": 3}
assert input_exp.evaluate(ctx) == output_exp.evaluate(ctx)

File: libraries/website/docs/snippets/cas/overview/rules_factor_out.ts
import { DistributiveFactorOutRule } from "mathy/core/rules"
import { ExpressionParser } from "mathy/core/parser"

const input = "4x + 2x"
const output = "(4 + 2) * x"
const parser = new ExpressionParser()

const input_exp = parser.parse(input)
const output_exp = parser.parse(output)

// Verify that the rule transforms the tree as expected
const change = new DistributiveFactorOutRule().apply_to(input_exp)
if (`${change.result}` !== output) {
  throw new Error("invalid output")
}

// Verify that both tress evaluate to the same value
const ctx = { x: 3 }
if (input_exp.evaluate(ctx) !== output_exp.evaluate(ctx)) {
  throw new Error("invalid result")
}
